action/protocol/account/util/util.go

```diff
@@ -79,21 +81,34 @@ func Recorded(sr protocol.StateReader, addr address.Address) (bool, error) {
 }
 
 // AccountState returns the confirmed account state on the chain
-func AccountState(sr protocol.StateReader, addr address.Address) (*state.Account, error) {
-	a, _, err := AccountStateWithHeight(sr, addr)
+func AccountState(ctx context.Context, sr protocol.StateReader, addr address.Address) (*state.Account, error) {
+	a, _, err := AccountStateWithHeight(ctx, sr, addr)
 	return a, err
 }
 
 // AccountStateWithHeight returns the confirmed account state on the chain with what height the state is read from.
-func AccountStateWithHeight(sr protocol.StateReader, addr address.Address) (*state.Account, uint64, error) {
+func AccountStateWithHeight(ctx context.Context, sr protocol.StateReader, addr address.Address) (*state.Account, uint64, error) {
 	pkHash := hash.BytesToHash160(addr.Bytes())
-	account := state.NewEmptyAccount()
+	account := &state.Account{}
 	h, err := sr.State(account, protocol.LegacyKeyOption(pkHash))
 	switch errors.Cause(err) {
 	case nil:
-		fallthrough
-	case state.ErrStateNotExist:
 		return account, h, nil
+	case state.ErrStateNotExist:
+		tip, err := sr.Height()
+		if err != nil {
+			return nil, 0, err
+		}
+		ctx = protocol.WithBlockCtx(ctx, protocol.BlockCtx{
+			BlockHeight: tip + 1,
+		})
+		ctx = protocol.WithFeatureCtx(ctx)
+		var opts []state.AccountCreationOption
+		if protocol.MustGetFeatureCtx(ctx).CreateLegacyNonceAccount {
+			opts = append(opts, state.LegacyNonceAccountTypeOption())
+		}
+		account, err = state.NewAccount(opts...)
+		return account, h, err
```

Comment:
also, the codebase has about 10 places calling this `AccountState()`, i think a better option is to create new account by the caller, like
```
state, err := accountutil.AccountState(sr, addr)
if err == ErrStateNotExist {
    // create new account (possibly with ZeroNonceOption) here
   state = NewAccount(opts...)
}
```
this way we don't need to touch these existing code, it will just be like 10 new changes on top of existing PR